proc/groupby/groupby.go

```diff
@@ -154,95 +155,143 @@ func New(pctx *proc.Context, parent proc.Interface, keys []expr.Assignment, aggN
 		parent:   parent,
 		agg:      agg,
 		resultCh: make(chan proc.Result),
+		doneCh:   make(chan struct{}),
 	}, nil
 }
 
-func (p *Proc) Pull() (zbuf.Batch, error) {
+func (p *Proc) Pull(done bool) (zbuf.Batch, error) {
+	if done {
+		for {
+			select {
+			case p.doneCh <- struct{}{}:
+				return nil, nil
+			case <-p.pctx.Done():
+				return nil, p.pctx.Err()
+			}
+		}
+	}
 	p.once.Do(func() { go p.run() })
 	if r, ok := <-p.resultCh; ok {
 		return r.Batch, r.Err
 	}
 	return nil, p.pctx.Err()
 }
 
-func (p *Proc) Done() {
-	p.parent.Done()
-}
-
 func (p *Proc) run() {
-	sendResults := func(p *Proc) error {
+	sendResults := func(p *Proc) bool {
 		for {
-			b, err := p.agg.Results(true, p.ectx)
-			if err != nil {
-				return err
+			b, err := p.agg.nextResult(true, p.batch)
+			done, ok := p.sendResult(b, err)
+			if !ok {
+				return false
 			}
-			if b == nil {
-				return nil
+			if b == nil || done {
+				return true
 			}
-			p.sendResult(b, nil)
 		}
 	}
-	var eof bool
+	defer func() {
+		close(p.resultCh)
+	}()
 	for {
-		batch, err := p.parent.Pull()
-		if err != nil || (batch == nil && eof) {
-			p.shutdown(err)
-			return
+		batch, err := p.parent.Pull(false)
+		if err != nil {
+			if _, ok := p.sendResult(nil, err); !ok {
+				return
+			}
+			continue
 		}
 		if batch == nil {
-			if err := sendResults(p); err != nil {
-				p.shutdown(err)
+			if ok := sendResults(p); !ok {
 				return
 			}
-			eof = true
+			if p.batch != nil {
+				p.batch.Unref()
+				p.batch = nil
+			}
 			continue
 		}
-		ectx := batch.Context()
-		p.ectx = ectx
-		eof = false
+		if p.batch == nil {
+			batch.Ref()
+			p.batch = batch
+		}
 		vals := batch.Values()
 		for i := range vals {
-			p.agg.Consume(ectx, &vals[i])
+			p.agg.Consume(batch, &vals[i])
 		}
 		batch.Unref()
 		if p.agg.inputDir == 0 {
 			continue
 		}
 		// sorted input: see if we have any completed keys we can emit.
 		for {
-			res, err := p.agg.Results(false, ectx)
+			res, err := p.agg.nextResult(false, batch)
 			if err != nil {
-				p.shutdown(err)
-				return
+				if _, ok := p.sendResult(nil, err); !ok {
+					return
+				}
+				break
 			}
 			if res == nil {
 				break
 			}
 			expr.SortStable(res.Values(), p.agg.keyCompare)
-			p.sendResult(res, nil)
+			done, ok := p.sendResult(res, nil)
+			if !ok {
+				return
+			}
+			if done {
+				break
+			}
 		}
 	}
 }
 
-func (p *Proc) sendResult(b zbuf.Batch, err error) {
+func (p *Proc) sendResult(b zbuf.Batch, err error) (bool, bool) {
 	select {
 	case p.resultCh <- proc.Result{Batch: b, Err: err}:
+		return false, true
+	case <-p.doneCh:
+		if b != nil {
+			b.Unref()
+		}
+		p.reset()
+		b, pullErr := p.parent.Pull(true)
+		if pullErr == nil {
+			err = pullErr
+		}
+		if err != nil {
+			select {
+			case p.resultCh <- proc.Result{Err: err}:
+				return true, false
+			case <-p.pctx.Done():
+				return false, false
+			}
+		}
```

Comment:
This doesn't feel right. If `err != nil` on the way in, shouldn't we send it on `p.resultCh` regardless of the value of `pullErr`?